collector/accelerators.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2024 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type acceleratorsCollector struct {
+	pciDevicesPath string
+	logger         log.Logger
+}
+
+func init() {
+	registerCollector("accelerator", defaultEnabled, NewAcceleratorCollector)
+}
+
+// NewAcceleratorCollector returns a new Collector exposing accelerator cards count.
+func NewAcceleratorCollector(logger log.Logger) (Collector, error) {
+	return &acceleratorsCollector{
+		pciDevicesPath: filepath.Join(*sysPath, "bus/pci/devices"),
+		logger:         logger,
+	}, nil
+}
+
+var (
+	acceleratorCardsDesc = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "accelerator", "cards"),
+		"Number of accelerator cards per type.",
+		[]string{"type"}, nil,
+	)
+
+	// vendor map, add any new vendor to this map
+	vendorToDeviceMap = map[string]map[string]string{
+		// nvidia devices
+		"0x10de": nvidiaDeviceIDsMap,
+	}
+)
+
+func (a *acceleratorsCollector) Update(ch chan<- prometheus.Metric) error {
+	devicesCount := map[string]int{}
+	pciDevices, err := os.ReadDir(a.pciDevicesPath)
+	if err != nil {
+		return fmt.Errorf("failed to read from  %q: %w", a.pciDevicesPath, err)
+	}
+
+	for _, pciDevice := range pciDevices {
+		vendorID, deviceID, err := a.getVendorAndDeviceIDs(pciDevice)
+		if err != nil {
+			level.Error(a.logger).Log("msg", "failed to get pci vendor and device IDs", "name", pciDevice.Name(), "err", err)
+			continue
+		}
+
+		level.Debug(a.logger).Log("msg", "checking pci device", "vendor", vendorID, "device", deviceID)
+
+		deviceDesc, isMonitored := isMonitoredAccelerator(vendorID, deviceID)
+		if !isMonitored {
+			continue
+		}
+		level.Debug(a.logger).Log("msg", "accelerator device found", "vendor", vendorID, "device", deviceID)
+		devicesCount[deviceDesc] = devicesCount[deviceDesc] + 1
+	}
+
+	for deviceDesc, count := range devicesCount {
+		ch <- prometheus.MustNewConstMetric(acceleratorCardsDesc, prometheus.CounterValue, float64(count), deviceDesc)
+	}
```

Comment:
(nit) let's sort the metrics by key name.
```suggestion
	keys := make([]string, 0, len(devicesCount))
	for k := range devicesCount {
		keys = append(keys, k)
	}
	slices.Sort(keys)

	for _, k := range keys {
		ch <- prometheus.MustNewConstMetric(acceleratorCardsDesc, prometheus.CounterValue, float64(devicesCount[k), k)
	}
```

Reply:
@simonpasquier what is the advantage of creating metrics in that particular order?

Reply:
It's cleaner if the metric payload doesn't vary between requests (it may be that the Prometheus client does this already but I'm not 100% sure).